allow for overlapping beatlooprolls

iamcodemaker opened the pull request:
This is necessary if multiple beatlooproll keys are pressed on a controller. Previously, Mixxx would stop the loop roll if any keys were released. Now it tracks all active rolls and will activate the last active one that is still active (button still down) instead of deactivating all rolls. Additionally, if a loop roll is active, activating another beat loop roll will start the loop from the start of the current loop, instead the current play position.

I believe this loop roll behavior is more intuitive and natural.

iamcodemaker committed:
allow for overlapping beatlooprolls

This is necessary if multiple beatlooproll keys are pressed on a
controller. Previously, Mixxx would stop the loop roll if any keys were
released. Now it tracks all active rolls and will activate the last
active one that is still active (button still down) instead of
deactivating all rolls. Additionally, if a loop roll is active,
activating another beat loop roll will start the loop from the start of
the current loop, instead the current play position.

daschuer commented:
Thank you for the Pull request.
Do we have a Launchpad bug for it?

daschuer reviewed (changes requested):
Works good, I have only some minor comments. 
Thank you.

iamcodemaker commented on src/engine/loopingcontrol.cpp:
```diff
@@ -846,10 +847,11 @@ void LoopingControl::slotBeatLoopActivateRoll(BeatLoopingControl* pBeatLoopContr
-    slotBeatLoop(pBeatLoopControl->getSize(), false, true);
+    slotBeatLoop(pBeatLoopControl->getSize(), m_bLoopRollActive, true);
     m_bLoopRollActive = true;
+    m_activeLoopRolls.push(pBeatLoopControl);
```
Yes. I'll make that change.

iamcodemaker replied:
If we store `double`s in here, are we going to have any issues comparing them using `==`? I don't understand floats well enough to say. The values should be exact though, so my gut says this is fine.

iamcodemaker commented on src/engine/loopingcontrol.cpp:
```diff
@@ -858,14 +860,21 @@ void LoopingControl::slotBeatLoopDeactivate(BeatLoopingControl* pBeatLoopControl
-    Q_UNUSED(pBeatLoopControl);
-    setLoopingEnabled(false);
+    pBeatLoopControl->deactivate();
+    m_activeLoopRolls.removeAll(pBeatLoopControl);
```
Will update. In theory there should only be one of each beat loop in here max. Actually thinking about it, if a controller had two ways to activate a loop roll we might end up with duplicates. So looping here is necessary.

iamcodemaker replied:
Made the removal loop O(n) (though the stack will always be so small it won't matter). It is slightly more complex now though.

iamcodemaker committed:
store loop sizes instead of pointers in m_activeLoopRolls

ronso0 commented:
Ooouh, sounds interesting!! I got used to the way it works right now, but your idea makes sense. Didn't test this yet.


Is this the right place to ask for taking a look at [lp:1721881 beatloop shouldn't alter beatloop size in spinbox](https://bugs.launchpad.net/mixxx/+bug/1721881) ?


Also, just yesterday I wished I could transfrom a **rolling** beatlopp into a 'regular' one, to resize it and move it. I got stucked when thinking about the controller implementation...

rryan commented:
Thanks! To make sure we dont break this in the future could you please add some unit tests?
src/test/looping_control_test.cpp

iamcodemaker commented:
Not sure if there is a launchpad bug for it, I didn't check. I'll work on some tests.

iamcodemaker committed:
add unit tests for new beatlooproll behavior

Tests validate the following:

- beatlooproll activation and deactivation work
- overlapping beatlooprolls remain active until the last roll is
  released
- overlapping beatlooprolls are properly activated as they stack and
  unwind
- new beatlooprolls start at the play position
- beatlooprolls activated while existing beatloops are in progress
  preserve the start point

daschuer commented:
@rons0 transforming a loop, in a way that you can release you finger and the loop keeps looping?
Thats a valid use case. Can you file a bug that we can consider details there?

ronso0 commented:
[lp:1793520 transform a rolling beatloop into a 'static' loop](https://bugs.launchpad.net/mixxx/+bug/1793520)

daschuer commented:
The Travis failure is unrelated.
LGTM Thank you very much.
